mineral-ui/mineral-ui
# [colors][theme] Update color palette & rework theme structure

### Description
- Update color palette per spec
- Remove `a11y_text_light` property from color objects
- Remove `getReadableTextColor`
- Remove `lime`, `orange`, and `yellow` colors; add `bronze`
- Update createTheme
    - Update with new/removed colors
    - Update border colors per spec
    - Change key naming scheme from `[property]_[target]_[variant]_[state]` to `[target]_[property]_[variant]_[state]`
    - [Rename lots of keys](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8)
    - Add "panel" and "well" targets
    - Add "inverted" concept
    - Add "icon" target
    - Add `backgroundColor_[state]`, `backgroundColor_theme_selected[state]`, `backgroundColor_themePrimary_[state]`, `backgroundColor_[variant]_hover
    - Add `color_readOnly` & `color_required`
    - Take keys out of `createTheme` into a new `tokens` and group them by target
- Change Icon's default fill to `'currentcolor'` to enable styling via `color` property
- Update components accordingly
    - Avatar
        - Update icon styles
        - Remove palette colors from `background` & `color` props
    - Button
        - Add `ButtonIcon_color` theme variable
        - Normalize focus styles with other components
        - Update icon styles
        - Remove/rename theme variables ([details](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8#button))
    - Choice (Checkbox, Radio)
        - Normalize focus styles with other components
        - Remove/rename theme variables ([details](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8#checkbox))
    - FauxControl (Select, TextArea, TextInput)
        - Add `Select_color_readOnly` theme variable (also applies to TextArea & TextInput)
        - Normalize focus styles with other components
        - Update icon styles
        - Remove/rename theme variables ([details](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8#formfield))
    - Menu
        - Add appropriate hover styles with a variant
        - Update icon styles
        - Remove/rename theme variables ([details](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8#menuitem))
    - Card, Dropdown, Form, Link, Popover, Text, Tooltip
        - Minor cleanup
        - Update icon styles
- Various component updates
    - Avatar: gray background uses text color, not simply black
    - Button, Choice: update to use `color_text`
    - Button, Select, TextArea, TextInput: normalize focus boxShadow styles
    - Card: change "borderColor" on focus
        - `color_theme_100` -> `borderColor_focus`
    - CardFooter: update to use `borderColor`; update icon color
    - CardTitleMenu: update icon color
    - Checkbox, Radio: update to use `backgroundColor_input`
    - Select, TextArea, & TextInput: correct input text color
    - Tooltip: ensure inline Tooltip triggers inherit font color
    - Apply `placeholder_color` to appropriate components
- Update color page, palette demo, and Avatar demo in website
    - Add requirement for "color_60" contrast ratio against white
    - Add `magenta` to selectable themes in palette demo
    - Update theme variable usage
    - Add sub-headings for targets in the theme structure table
- Update relevant specs & snapshots

> **Note for reviewers**: Given the amount of visual changes, I've left the happo report unresolved. Please review in detail and approve or reject as you see fit. FWIW, I have already done so and found no issues.

### Motivation and context


Closes #570, closes #553, closes #555, closes #558, closes #560, closes #562, closes #564

### Screenshots, videos, or demo, if appropriate


https://570-revise-color-palette--mineral-ui.netlify.com/color

### How to test


1. `cd mineral-ui && npm start`
1. Navigate to [color page](https://570-revise-color-palette--mineral-ui.netlify.com/color)
1. Confirm color ramps look correct per spec
1. Navigate to [palette demo](https://570-revise-color-palette--mineral-ui.netlify.com/palette-demo)
1. Confirm the new magenta theme and all other themes work correctly
1. Navigate to [Avatar demo](https://570-revise-color-palette--mineral-ui.netlify.com/components/avatar)
1. Confirm the colors example uses all white text (except for the gray and custom ones)
1. Check all component examples and all site pages for correctness
    1. This PR touches a _lot_, so it's good to be thorough. I did my best throughout this process, but it would still benefit from as many eyes as we can put on it.
1. Review the [changelog-to-be](https://gist.github.com/kylegach/00a2b9aeffadf47e54accd858d158249)
    1. Because there's so many breaking changes, communicating those changes is extra important.

### Types of changes

- Breaking change (fix or feature that would cause existing functionality to change)
- Other (provide details below)

#### Breaking
- **colors**: Remove `lime`, `orange`, and `yellow`
- **Avatar**: Remove lime, orange, & yellow from accepted background colors
- **Icon**: Change default fill to `'currentcolor'` from `theme.color_gray_60`
- See [this gist](https://gist.github.com/kylegach/3f1300c071c8b47952eb12729e5a90e8) for a detailed changelog, re: theme variables, both base and component-level
- Here's a [preview of the changelog-to-be](https://gist.github.com/kylegach/00a2b9aeffadf47e54accd858d158249)

### Checklist




### How does this PR make you feel?

![Colorful fractal](https://media.giphy.com/media/wcgcRnkePIUPS/giphy.gif)
